lineapy/graph_reader/session_artifacts.py

```diff
@@ -264,24 +264,40 @@ def _update_node_context(self):
                         len(self.node_context[node_id].dependent_variables)
                         == 0
                     )
-                    has_assigned_before = (
-                        self.input_parameters_node.get(var, None) is not None
+                    previous_assignment = self.input_parameters_node.get(
+                        var, None
                     )
-                    if is_literal_assignment:
-                        if has_assigned_before:
+
+                    if previous_assignment is None:
+                        self.input_parameters_node[var] = node_id
+                    else:
+                        previous_assignment_is_literal = (
+                            len(
+                                self.node_context[
+                                    previous_assignment
+                                ].dependent_variables
+                            )
+                            == 0
+                        )
+
+                        if (
+                            is_literal_assignment
+                            and previous_assignment_is_literal
+                        ):
                             raise ValueError(
-                                "Variable %s, is defined more than once", var
+                                f"Variable {var}, is defined more than once"
                             )
-                        else:
+                        elif not previous_assignment_is_literal:
+                            # previous assignment is not literal, so we can reassign it
                             self.input_parameters_node[var] = node_id
+                        # else previous assignment is literal and we should not override it
```

Comment:
This looks like a feature but might bite us in the future when we start supporting non-literal assigns. Non-literal assigns I think are more important than primitives and the user might very well mean the non-literal assign (and instead get the mistakenly named string as a variable instead)

That said if it's a solution to a particular problem then disregard this comment.

Reply:
When we support non-literal assigns this whole block needs to change to disallow any duplicates regardless of if they are non-literals or not. 

If we decide that we want to disallow any duplicates (even non-literals we don't support now) this code needs to be struck out to disallow and reassignments.

Reply:
Keeping as is, added comment to clarify what case this is meant to catch before this block.